dev/PushNotifications/PushNotificationReceivedEventArgs.cpp

```diff
@@ -0,0 +1,118 @@
+﻿#include "pch.h"
+
+#include <winrt/Windows.ApplicationModel.background.h>
+#include <winrt/Windows.Storage.Streams.h>
+#include <winrt/Windows.Networking.PushNotifications.h>
+
+#include "PushNotificationReceivedEventArgs.h"
+#include "Microsoft.Windows.PushNotifications.PushNotificationReceivedEventArgs.g.cpp"
+#include <iostream>
+
+namespace winrt
+{
+    using namespace Windows::ApplicationModel::Background;
+    using namespace Windows::Storage::Streams;
+    using namespace Windows::Networking::PushNotifications;
+}
+
+namespace winrt::Microsoft::Windows::PushNotifications::implementation
+{
+    PushNotificationReceivedEventArgs::PushNotificationReceivedEventArgs(winrt::IBackgroundTaskInstance const& backgroundTask)
+    {
+        m_backgroundTaskInstance = backgroundTask;
+
+        // Need to process the RawNotification payload here because TriggerDetails is set to nullptr leaving IBackgroundTaskInstance::Run 
+        auto triggerDetails = m_backgroundTaskInstance.TriggerDetails();
+        RawNotification rawNotification = triggerDetails.as<RawNotification>();
+
+        IBuffer rawContentAsBuffer = rawNotification.ContentBytes();
+        DataReader dataReader = DataReader::FromBuffer(rawContentAsBuffer);
+
+        m_backgroundPayload = winrt::com_array<uint8_t>(rawContentAsBuffer.Length());
+        dataReader.ReadBytes(m_backgroundPayload);
+    }
+
+    PushNotificationReceivedEventArgs::PushNotificationReceivedEventArgs(winrt::PushNotificationReceivedEventArgs const& args)
+    {
+        m_args = args;
+    }
+
+    winrt::Microsoft::Windows::PushNotifications::PushNotificationReceivedEventArgs PushNotificationReceivedEventArgs::CreateFromBackgroundTaskInstance(winrt::IBackgroundTaskInstance const& backgroundTask)
+    {
+        return winrt::make<PushNotificationReceivedEventArgs>(backgroundTask);
+    }
+
+    winrt::Microsoft::Windows::PushNotifications::PushNotificationReceivedEventArgs PushNotificationReceivedEventArgs::CreateFromPushNotificationReceivedEventArgs(winrt::PushNotificationReceivedEventArgs const& args)
+    {
+        return winrt::make<PushNotificationReceivedEventArgs>(args);
+    }
+
+    winrt::com_array<uint8_t> PushNotificationReceivedEventArgs::Payload()
+    {
+        auto lock = m_lock.lock_shared();
+        winrt::com_array<uint8_t> payload{};
+        if (m_args)
```

Comment:
Why do we get the payload on-demand for the args case, but get it up front for the task case? Either always get on demand or always get up front.

For always-on-demand, it would be

```cpp
return GetRawPayload(m_rawNotification);
```

where `m_rawNotification` is initialized in the constructor either from `m_args.RawNotification()` or from `backgroundTask.TriggerDetails().as<RawNotification>()`.

For up-front, we would extract the raw payload in the constructor.

I think on-demand is probably better, assuming most people aren't going to ask for the raw payload. There's not much point caching the raw payload, given that it's already being cached inside the `RawNotification` object. Don't cache something that's already cached.

Reply:
Sadly, we have to process the payload up front for the background task case because TriggerDetails gets deallocated when it leaves Run. This isn't the case for args so that's why there is a up-front/on-demand.

If we want to keep the implementation of both the we can switch to one m_rawNotification var that is processed up-front in the constructor. What do you suggest here?

Reply:
I would convert to `m_rawNotification` in the constructor. That way, the member can be const.